document/core/syntax/types.rst

```diff
@@ -73,39 +73,84 @@ Conventions
 
 
 
-.. index:: ! heap type, store, type identifier, ! substitution, ! closed type
+.. index:: ! heap type, store, type identifier, ! substitution, ! closed type, ! abstract type, ! concrete type, ! unboxed scalar
    pair: abstract syntax; heap type
+.. _type-i31:
 .. _type-subst:
 .. _type-closed:
+.. _type-abstract:
+.. _type-concrete:
 .. _syntax-heaptype:
 
 Heap Types
 ~~~~~~~~~~
 
 *Heap types* classify objects in the runtime :ref:`store <store>`.
+There are three disjoint hierarchies of heap types:
+
+- *function types* classify :ref:`functions <syntax-func>`,
+- *aggregate types* classify dynamically allocated *managed* data, such as *structures*, *arrays*, or *unboxed scalars*,
+- *external types* classify *external* references possibly owned by the :ref:`embedder <embedder>`.
+
+The values from the latter two hierarchies are interconvertible by ways of the |EXTERNINTERNALIZE| and |EXTERNINTERNALIZE| instructions.
+That is, both type hierarchies are inhabited by an isomorphic set of values, but may have different, incompatible representations in practice.
 
 .. math::
    \begin{array}{llll}
    \production{heap type} & \heaptype &::=&
-     \FUNC ~|~ \EXTERN ~|~ \typeidx ~|~ \functype ~|~ \BOT \\
+     \FUNC ~|~ \NOFUNC \\&&|&
+     \EXTERN ~|~ \NOEXTERN \\&&|&
+     \ANY ~|~ \EQ ~|~ \I31 ~|~ \STRUCT ~|~ \ARRAY ~|~ \NONE \\&&|&
+     \typeidx ~|~ \deftype ~|~ \BOT \\
    \end{array}
 
-The type |FUNC| denotes the infinite union of all types of :ref:`functions <syntax-func>`, regardless of their concrete :ref:`function types <syntax-functype>`.
+A heap type is either *abstract* or *concrete*.
+
+The abstract type |FUNC| denotes the infinite union (i.e., supertype) of all :ref:`function types <syntax-functype>`, regardless of their concrete definition.
+Dually, the type |NOFUNC| denotes the infinite intersection (i.e., subtype) of all :ref:`function types <syntax-functype>`, regardless of their concrete definition.
+This type has no values.
+
+The abstract type |EXTERN| denotes the infinite union of all external references received through the :ref:`embedder <embedder>`.
+This type has no concrete subtypes.
+Dually, the type |NOEXTERN| denotes the infinite intersection of all forms of external references.
+This type has no values.
+
+The abstract type |ANY| denotes the infinite union of all aggregate types, as well as possibly abstract values produced by *internalizing* an external reference of type |EXTERN|.
+Dually, the type |NONE| denotes the infinite intersection of all forms of aggregate types.
+This type has no values.
+
+The abstract type |EQ| is a subtype of |ANY| that includes all types for which references can be compared, i.e., aggregate values and |I31|.
+
+The abstract types |STRUCT| and |ARRAY| denote the infinite union of all :ref:`structure <syntax-structtype>` and :ref:`array <syntax-arraytype>` aggregates, respectively.
+
+The abstract type |I31| denotes *unboxed scalars*, that is, integers injected into references.
```

Comment:
Is i31 really abstract? It seems that it should be considered concrete since there are values for which it is the dynamic type. Maybe instead of "concrete" we should be using "defined" or similar?

Reply:
Maybe "declared" and "undeclared"?  "undeclared" seems accurate to describe all the built-in types that are not declared/defined in the type section, but idk if we have precedent talking about declarations at all.

Reply:
Right, I think we don't currently talk about "declarations" anywhere. But more importantly, "undeclared" still has this strong connotation of being erroneous, doesn't it?

Reply:
Hmm maybe, although not nearly as much as "undefined" IMO. Maybe "declared" and "built in" would work?

Anyway, I don't think figuring out the perfect names here should block landing this.